docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+  generation numbers.
+
+The node generation number defines a global "in" definition for pageserver nodes: a node whose
+generation matches the generation the control plane most recently issued is a member of the cluster
+in good standing. Any node with any older generation ID is not.
+
+#### Why use per-tenant _and_ per-node generation numbers?
+
+The essential generation number is the tenant attachment number: this alone would be sufficient
+to implement safe migration and failover. However, it would also require that all such events
+carry an O(tenant_count) cost, or to have a global sequence number for all changes to attachments (which
+would imply synchronization between all attachment changes): the pageserver generation number
+is included to avoid writing this kind of system design constraint into the storage format.
+
+The pageserver generation number acts as an optimization and also provides flexibility in future designs:
+
+- if we wanted to have pageservers do their own HA in future without control plane in the loop on a per-tenant basis,
+  then node generation numbers make that possible.
+- the pageserver generation number also enables future storage of per-pageserver data in S3 (e.g.
+  housekeeping type content) that may not be within the scope of a tenant attachment.
+- pageserver generation numbers enable building fencing into any network protocol (for example
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+The tenant attachment generation and node generation are mostly independent, but have one key
+coupling:
+
+- Control plane must not "steal" a tenant from a node (by attaching to another node without detaching
+  from the first node) without also incrementing the original node's generation number, which
+  effectively marks the original node as dead.
+- This enables nodes to trust that if their generation ID is still current, then all their
+  attachments are as well.
+
+### Object Key Changes
+
+All object keys (layer objects and index objects) will have a suffix as follows:
+
+```
+   <node id>-<attachment generation>-<node generation>
+```
+
+To avoid unreasonably long keys, we may make some reasonable assumptions about max lengths
+of these integers:
+
+- 32 bit attachment generation (this easily accommodates reasonable worst case scenarios like
+  a node restarting every minute for 5 years)
+- 16 bit node ID (this doesn't have to accommodate all pageservers ever, just all the pageservers
+  that act together in a logical cluster)
+- 32 bit node generation (as with attachment generation, this can handle even very frequent restarts)
+
+So an example suffix for generation 1 of node 0 might look like:
+
+```
+   00000001-0000-00000001
+```
+
+This suffix provides a lot of flexibility in how we may handle migration/restarts/failover
+at runtime:
+
+- We can replace a pageserver with another physical node having the same pageserver ID, and
```

Comment:
Added sentence: 
> This applies even if we have no robust way to terminate a partitioned pageserver: the older-generation node may  continue writing objects under their own generation number, but those will not conflict with objects written by the newer generation pageserver.

The key idea is that it becomes safe to leave a "rogue" pageserver running in the scenario where we have no robust way to kill it (e.g. EC2 API takes too long to terminate an instance).  In practice this would be rare, and the activity of the rogue node would be bounded by places it has to touch other services (see "Deletion changes" where it has to validate its generation before executing deletes, or the Safekeeper section where it is proposed to make safekeepers generation-aware and refuse requests from such rogue nodes).

Reply:
This conversation has been addressed well in the current form of the RFC, closing it.